Use error codes in plutus scripts

ffakenz opened the pull request:
## Why

Plutus error messages respect no particular convention. We just write the error message how we feel it when we write the code.

Plutus error messages are checked by the mutation test. So we duplicate constant string between production code and test code to check that a specific error message has been emitted by the validator.

Plutus error messages, in my experience, are only used to find, in the code, where the error has been generated and understand what went wrong by looking at this code and not, just, the message per se. That means that we could use unique error codes instead of plain text strings and get the same value. The benefit would be to significantly reduce our scripts size as these plain text strings are way bigger than what we could do with error codes.

## What

TBD: I suggest we have some constants somewhere whose name are human readable and whose value are error codes. We use template Haskell to insert the error code from the variable in the plutus script. We use the variable in the mutation to assert for the appropriate error code to have been emitted.

## How

Discussed in slack https://input-output-rnd.slack.com/archives/C01PKL5P6PJ/p1674026194666629?thread_ts=1674024828.046979&cid=C01PKL5P6PJ


<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [ ] Documentation updated
* [ ] Added and/or updated haddocks
* [ ] No new TODOs introduced or explained herafter

ffakenz committed:
Replace trace error messages by error codes

This should make scripts smaller and with helper function 'toErrorCode'
in we can keep errors meaningful in mutation tests.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -835,3 +835,178 @@ replaceContesters contesters = \case
+    PartySignatureVerificationFailed -> "H36"
+
+toErrorCode :: ToErrorCode a => a -> Text
+toErrorCode = toCode
```
I would move these things into the `hydra-plutus` package into the respective script modules even though we don't use them from plutus-tx. For eample `HeadError` into `Hydra.Contract.Head` module. The closer we keep the strings together, the more likely they stay consistent

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -63,7 +63,7 @@ import qualified PlutusTx.Builtins as Builtins
 type RedeemerType = Input
 
---------------------------------------------------------------------------------
+-------------------------------------------------------------------------------- -> ""
```
Remove this random change

ffakenz replied:
my bad, good catch!

v0d1ch commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -835,3 +835,178 @@ replaceContesters contesters = \case
+    STIsMissingInTheOutput -> "C02"
+
+instance ToErrorCode HeadError where
+  toCode :: HeadError -> Text
```
Potentially we could generate these instances using TH or use functions from `Data.Data`

ffakenz replied:
Using TH was the idea here, but I do not know how, so we decided to open this PR in a dirty way first.

Do you think we should do it in this PR?

ffakenz replied:
From our tactical meeting, we agreed that we will tackle this in a different PR.

ffakenz committed:
Remove typo

ffakenz committed:
Factor out error codes to their respective script modules in order to favor consistency

ffakenz committed:
Increase the max number of parties supported from 3 back to 4

ffakenz committed:
Increase the max number of parties supported from 4 to 6

ch1bo committed:
Maximum number of parties supports is 4

The StateSpec tests do not check the execution budgets, but the tx-cost
benchmark does. However, it was not always trying to abort all commits -
this is changed in this commit now.

In the mid-term, we should fix the property in StateSpec to check
budgets as well or change the claim.

ffakenz committed:
Remove redundant import

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -223,13 +223,13 @@ genContestMutation
     ) =
     oneof
-      [ SomeMutation Nothing MutateSignatureButNotSnapshotNumber . ChangeHeadRedeemer <$> do
+      [ SomeMutation (Just "signature verification failed") MutateSignatureButNotSnapshotNumber . ChangeHeadRedeemer <$> do
```
This commit and the next should not be in this PR, right? There are no error codes used in this one and I think you might have mixed this up with https://github.com/input-output-hk/hydra/pull/719?

ch1bo replied:
@ffakenz Seems like these came from a rebase problem. Can you please just remove the commits which ended up wrong here (it's useless history).

ch1bo replied:
I did this now and updated your `mutations/explicit_expected_errors_2` branch with these two commits (squashed)

ffakenz replied:
Yeah, my bad had an inconvenience yesterday and by mistake, force-pushed the commits into the wrong branch.

Nice catch and thank you for the help.

ch1bo committed:
Move ToErrorCode into dedicated module

Also simplify definition to directly use and define toErrorCode.

Why? Small modules are good and Util is usually becoming a kitchen sink
quite fast.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Error.hs:
```diff
@@ -0,0 +1,8 @@
+import Hydra.Prelude (Show, Text)
+
+-- | Types which are used to describe errors as short error codes in scripts.
+class Show a => ToErrorCode a where
```
`Show` is actually not needed in this PR. We should not introduce this constraint right now.

ffakenz committed:
Remove unnecessary Show superclass constraint for ToErrorCode
